### PR TITLE
Event all the things!

### LINKED ISSUE #13845: Proposal: names and labels in events api
It would be good to have the events api endpoint provide the names and labels with the events. For people than encode metadata in the name/label. Otherwise I'd need to do an inspect after receiving an event or have a cache of id(s) to metadata
It looks fairly straightforward to implement  and I can send a PR.

docker version:
docker info:
uname -a:

Provide additional environment details (AWS, VirtualBox, physical, etc.):
List the steps to reproduce the issue:
Describe the results you received:
Describe the results you expected:
Provide additional info you think is important:

### LINKED ISSUE #18164: Improve metadata of `/events` API endpoint
Follow-up issue after discussion with @calavera and @thaJeztah at dockercon.

At the moment, it's not trivial to see if an event is connected to container, image, volume or network:

```json
{"status":"create","id":"5745704abe9caa5","from":"busybox","time":1442421716,"timeNano":1442421716853979870}
{"status":"delete","id":"7f05242655562b48fec7dc838074d0023db80c69f750cb41ab2544596c015d49","time":1448281959}
```

It would be awesome if the event could contain object type, e.g.:

```
"type": "container"
```

or

```
"type": "volume"
```

I am planning to consume this in TUI for engine I wrote: https://github.com/TomasTomecek/sen

### LINKED ISSUE #18453: `docker events --filter container=foo` is broken as of 1.9.0
When running `docker events` with a `container` filter, nothing is received, even though events from a container with that name are received by the daemon. This is a regression from 1.8.3.

```
bjaglin@t440s:/tmp$ docker version
Client:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.3
 Git commit:   a34a1d5
 Built:        Fri Nov 20 17:56:04 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.3
 Git commit:   a34a1d5
 Built:        Fri Nov 20 17:56:04 UTC 2015
 OS/Arch:      linux/amd64
```
```
bjaglin@t440s:/tmp$ docker info
Containers: 1
Images: 14
Server Version: 1.9.1
Storage Driver: overlay
 Backing Filesystem: extfs
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.2.0-18-generic
Operating System: Ubuntu 15.10
CPUs: 4
Total Memory: 11.64 GiB
Name: t440s
ID: 4CO2:NGIJ:KC6L:MBYG:BXNH:DRJK:GM5J:LQHT:H6QD:7RA2:FX2O:NRXF
WARNING: No swap limit support
```
```
bjaglin@t440s:/tmp$ uname -a
Linux t440s 4.2.0-18-generic #22-Ubuntu SMP Fri Nov 6 18:25:50 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
```

Actual:
```
bjaglin@t440s:/tmp$ docker events --filter container=foo &
[1] 17134
bjaglin@t440s:/tmp$ docker run --rm --name foo alpine echo hello world
hello world
bjaglin@t440s:/tmp$ fg
docker events --filter container=foo
```

Expected (1.8.3 behavior):
```
bjaglin@t440s:/tmp$ docker events --filter container=foo &
[1] 30008
bjaglin@t440s:/tmp$ docker run --rm --name foo alpine echo hello world2015-12-06T00:24:18.000000000+01:00 9c683202e1fe738111261e9601421c31bfe53b7926f52bbcb941238da0a96cc8: (from alpine) create
2015-12-06T00:24:18.000000000+01:00 9c683202e1fe738111261e9601421c31bfe53b7926f52bbcb941238da0a96cc8: (from alpine) attach
2015-12-06T00:24:19.000000000+01:00 9c683202e1fe738111261e9601421c31bfe53b7926f52bbcb941238da0a96cc8: (from alpine) start
hello world
2015-12-06T00:24:19.000000000+01:00 9c683202e1fe738111261e9601421c31bfe53b7926f52bbcb941238da0a96cc8: (from alpine) die
bjaglin@t440s:/tmp$ fg
docker events --filter container=foo
```

Issue reproducible on 1.9.0, 1.9.1 & experimental (20151205).

### LINKED ISSUE #18672: Flaky test: TestStatsAllNewContainersAdded
From #18661 on janky

```
----------------------------------------------------------------------
FAIL: docker_cli_stats_test.go:94: DockerSuite.TestStatsAllNewContainersAdded

docker_cli_stats_test.go:125:
    c.Fatal("failed to observe new container created added to stats")
... Error: failed to observe new container created added to stats


----------------------------------------------------------------------
```

### LINKED ISSUE #18673: Flaky test: TestEventsImageImport
From #18661 and #18399 on Windows:

```
----------------------------------------------------------------------
FAIL: docker_cli_events_test.go:217: DockerSuite.TestEventsImageImport

docker_cli_events_test.go:254:
    c.Fatal("failed to observe image import in timely fashion")
... Error: failed to observe image import in timely fashion


----------------------------------------------------------------------
```

### LINKED ISSUE #18674: Flaky test: TestBuildCancellationKillsSleep
From #18661 on Windows:

```
----------------------------------------------------------------------
FAIL: docker_cli_build_test.go:1874: DockerSuite.TestBuildCancellationKillsSleep

docker_cli_build_test.go:1945:
    c.Fatal("failed to observe build container start in timely fashion")
... Error: failed to observe build container start in timely fashion


----------------------------------------------------------------------
```

### LINKED ISSUE #18711: Flaky test: TestEventsStreaming
From #18682 on Windows: https://jenkins.dockerproject.org/job/Windows-PRs/18783/console

```
----------------------------------------------------------------------
FAIL: docker_cli_events_test.go:422: DockerSuite.TestEventsStreaming

docker_cli_events_test.go:470:
    c.Fatal("failed to observe container create in timely fashion", "\n", buffer.String())
... Error: failed to observe container create in timely fashion
2015-12-16T00:48:55.061057000Z 3fad02e3f38327023fe272afcae5be0a9c5e41fd1caed5746691717937c18d2e: (from busybox) stop
2015-12-16T00:48:55.291022100Z 3fad02e3f38327023fe272afcae5be0a9c5e41fd1caed5746691717937c18d2e: (from busybox) destroy



----------------------------------------------------------------------
```

### PR DESCRIPTION
**Pull request by calavera**
This is an attempt to make our events more rich and extensible.

In order to do that, we stop using `JSONMessage` as a struct to deliver events. That struct is an abomination for all kind of purposes, and the less we use it the better (imo). I created a specific purpose structure that includes new fields:
- Type: it's the type of resource the event is triggered for, `container`, `image`, `volume` and `network`.
- Action: it's the name of the event itself, `create`, `destroy`, `mount`, `connect` and so on.
- Actor: it's information about the resource, `ID`, `Name` and `Attributes`. Attributes is an open map to send extra metadata, like the labels for a container event.

**IMPORTANT**: This change is backwards compatible at the API level, but not at the CLI level.

The new event structure preserves old fields that we've been using until now for container and image events. People consuming the API won't see any difference in those events, but will start to see events without information in those fields, because they come from storage and network operations.

I've included a test to guarantee that we keep that backwards compatible information: https://github.com/docker/docker/compare/master...calavera:event_types?expand=1#diff-1e68c88f5c9fa605c0f84199cd04feb0R38

The CLI output has been completely modified and it's standard for all types of events, the syntax is as follow:

```
timestamp actor_type action actor_id (<optional>actor_attributes)
```

The attributes are listed in KEY=VALUE format to be consistent with other places where we show this kind of values, like labels.
### Networking and container events

Network events that operate on containers (connect and disconnect) include the container id as part of the attributes:

```
2015-12-23T21:38:24.705709133Z network create 8b111217944ba0ba844a65b13efcd57dc494932ee2527577758f939315ba2c5b (name=test-event-network-local, type=bridge)
2015-12-23T21:38:25.075955890Z container create b4be644031a3d90b400f88ab3d4bdf4dc23adb250e696b6328b85441abe2c54e (name=test-network-container, image=busybox)
2015-12-23T21:38:25.119625123Z network connect 8b111217944ba0ba844a65b13efcd57dc494932ee2527577758f939315ba2c5b (name=test-event-network-local, container=b4be644031a3d90b400f88ab3d4bdf4dc23adb250e696b6328b85441abe2c54e, type=bridge)
2015-12-23T21:38:25.121051956Z container start b4be644031a3d90b400f88ab3d4bdf4dc23adb250e696b6328b85441abe2c54e (name=test-network-container, image=busybox)
```
### Storage and container events:

Storage events that operate on containers include the container id as part of the attributes:

```
2015-12-23T21:05:28.136212689Z volume create test-event-volume-local (driver=local)
2015-12-23T21:05:28.383462717Z volume mount test-event-volume-local (read/write=true, container=562fe10671e9273da25eed36cdce26159085ac7ee6707105fd534866340a5025, destination=/foo, driver=local, propagation=rprivate)
2015-12-23T21:05:28.650314265Z volume unmount test-event-volume-local (container=562fe10671e9273da25eed36cdce26159085ac7ee6707105fd534866340a5025, driver=local)
2015-12-23T21:05:28.716218405Z volume destroy test-event-volume-local (driver=local)
```
## Future improvements out of the scope here
1. Add a format flag with a go template.
## TODO
- [x] Modify the events documentation.

/cc @coolljt0725, @dnephin

- Fixes #18164
- Fixes #13845
- Fixes #18453 
- Fixes #18711
- Fixes #18674
- Fixes #18673
- Fixes #18672

### CODE REVIEW

**Comment by thaJeztah** [comment]
Nice! Looking at the command line output, should we put the event earlier on? the type of event is kinda hurried now.  E.g.

```
2015-12-23T21:38:24.705709133Z network create 8b111217944ba0ba844a65b13efcd57dc494932ee2527577758f939315ba2c5b (test-event-network-local) (type: bridge)
2015-12-23T21:38:25.075955890Z container create b4be644031a3d90b400f88ab3d4bdf4dc23adb250e696b6328b85441abe2c54e (test-network-container) (image: busybox)
2015-12-23T21:38:25.119625123Z network connect 8b111217944ba0ba844a65b13efcd57dc494932ee2527577758f939315ba2c5b (test-event-network-local) (container: b4be644031a3d90b400f88ab3d4bdf4dc23adb250e696b6328b85441abe2c54e, type: bridge)
2015-12-23T21:38:25.121051956Z container start b4be644031a3d90b400f88ab3d4bdf4dc23adb250e696b6328b85441abe2c54e (test-network-container)  (image: busybox)
```

Or 

```
2015-12-23T21:38:24.705709133Z create network 8b111217944ba0ba844a65b13efcd57dc494932ee2527577758f939315ba2c5b (test-event-network-local) (type: bridge)
2015-12-23T21:38:25.075955890Z create container  b4be644031a3d90b400f88ab3d4bdf4dc23adb250e696b6328b85441abe2c54e (test-network-container) (image: busybox)
2015-12-23T21:38:25.119625123Z connect network 8b111217944ba0ba844a65b13efcd57dc494932ee2527577758f939315ba2c5b (test-event-network-local) (container: b4be644031a3d90b400f88ab3d4bdf4dc23adb250e696b6328b85441abe2c54e, type: bridge)
2015-12-23T21:38:25.121051956Z start container b4be644031a3d90b400f88ab3d4bdf4dc23adb250e696b6328b85441abe2c54e (test-network-container)  (image: busybox)
```

We could even consider a `--format` flag in the future

**Comment by calavera** [comment]
Obligatory regular expression meme (do not look at `events_utils.go`):

image

@thaJeztah I'm okay with moving the action at the beginning, let's wait to hear from other people to see if we agree in a format and I only need to change it once :smile_cat:

**Comment by thaJeztah** [comment]
@calavera haha, absolutely, just my first impression looking at it. great work otherwise, I like it!

**Comment by thaJeztah** [comment]
Oh, and
/cc @TomasTomecek @samsabed

File: daemon/daemon.go [hunk 1]
```diff
@@ -1042,7 +1033,12 @@ func (daemon *Daemon) TagImage(newTag reference.Named, imageName string) error {
 	if err := daemon.referenceStore.AddTag(newTag, imageID, true); err != nil {
 		return err
 	}
-	daemon.EventsService.Log("tag", newTag.String(), "")
+
+	actor := eventtypes.Actor{
+		ID:   imageID.String(),
+		Name: newTag.String(),
+	}
+	daemon.EventsService.Log("tag", eventtypes.ImageEventType, actor)
```
Comment (dnephin) [inline daemon/daemon.go]: I believe all the `eventtypes.ImageEventType` events should include the image labels in `Actor.Attributes`. Previously this was being injected by `daemon.GetLabels()`

File: daemon/daemon.go [hunk 2]
```diff
@@ -549,20 +549,11 @@ func (daemon *Daemon) GetByName(name string) (*container.Container, error) {
...
-	// incoming container filter can be name, id or partial id, convert to
-	// a full container id
-	for _, cn := range filter.Get("container") {
-		c, err := daemon.GetContainer(cn)
-		filter.Del("container", cn)
-		if err == nil {
-			filter.Add("container", c.ID)
-		}
-	}
-	return events.NewFilter(filter, daemon.GetLabels)
```
Comment (dnephin) [inline daemon/daemon.go]: I think this method `getEventFilter()` and `daemon.GetLabels()` can both be delete now.

**Comment by dnephin** [comment]
Nice! I think this is awesome. We'll be able to make use of the new event data in the soon-be-released `docker-compose events` command.

File: daemon/daemon.go (hunk 1 above)
Comment (calavera) [inline daemon/daemon.go]: fixed, thanks!

File: daemon/daemon.go (hunk 2 above)
Comment (calavera) [inline daemon/daemon.go]: done! :boom:

File: daemon/daemon.go (hunk 1 above)
Comment (dnephin) [inline daemon/daemon.go]: I think they're all done except for this one place?

Comment (calavera) [inline daemon/daemon.go]: :pensive: fixing

Comment (calavera) [inline daemon/daemon.go]: actually there are a couple more places in pull/push :goberserk:

**Comment by calavera** [comment]
@thaJeztah one thing that I don't like about your suggested syntax is that it becomes harder to parse, because both name and attributes are optional at the end and are between parenthesis. But I'm open to suggestions.

**Comment by calavera** [comment]
> But I'm open to suggestions.

maybe we should add the name to the attributes. That would leave us only with one option field, which I think it's better:

```
timestamp type action id (<optional>attributes)
```

Example:
```
2015-12-23T21:38:25.075955890Z container create b4be644031a3d90b400f88ab3d4bdf4dc23adb250e696b6328b85441abe2c54e (name: test-network-container, image: busybox)
```

**Comment by TomasTomecek** [comment]
This is really impressive. Very nice work! A couple questions.

Will there be a precise documentation of what are the possible values (states) to expect? Regarding the new changes to layering, image IDs 'n stuff, would it make sense to also include digest? How about untagging an image, will I get the tag name? And last thing: would be super cool to unite events status and `inspect` status (`inspect` status contains text starting with capital letter and human-readable time which is a nightmare to parse).

**Comment by thaJeztah** [comment]
@calavera oh, I like the suggestion of adding name to attributes. In addition, we can still add `--format` in a future PR. My primary concern was that the "type" of event got a bit lost in the other information, putting it more to the start of the line makes that easier to read (visually).

**Comment by calavera** [comment]
> Regarding the new changes to layering, image IDs 'n stuff, would it make sense to also include digest? How about untagging an image, will I get the tag name?

@TomasTomecek yes, with the new attributes, we can include more relevant information about the events, like that one. This is not a one time only change, we can improve the information we add to each event in the future.

> And last thing: would be super cool to unite events status and inspect status (inspect status contains text starting with capital letter and human-readable time which is a nightmare to parse).

I'm not completely sure that's related to this change. I think we should discuss it further, but I understand your issues.

**Comment by calavera** [comment]
@thaJeztah I've changed the log format to put the name in the attributes.

I've also added a new feature to be able to filter by event type with `--filter type=eventType`, like `--filter type=container` or `--filter type=network`.

Can I have some :+1: / :-1: to all this before moving it to code review?

@TomasTomecek documentation is coming.

**Comment by vdemeester** [comment]
Really nice ! :+1: :wink:

**Comment by thaJeztah** [comment]
@calavera thanks!

- moving name to attributes: lgtm
- filter by "type": lgtm

One think I'm still thinking about is if "type" is the proper name for this. My first impression would be that "type" would be "start", "stop", "build". Not sure what we would use for "container" / "image" (source? origin? class?)

Perhaps someone else has suggestions here for a better naming.

**Comment by vdemeester** [comment]
@thaJeztah hum, I thought `type` was good because it follows the `type` of `inspect` though :stuck_out_tongue:.

**Comment by calavera** [comment]
> My first impression would be that "type" would be "start", "stop", "build". Not sure what we would use for "container" / "image" (source? origin? class?)

That could be an interpretation, I'm going to give you another one :smile: 

The information a event holds does not describe the event itself, it describes the change of behaviors an actor(something inside docker) is exposed to. `type` does NOT mean `event type`, although I've made the mistake of naming several variables `eventType` myself, and I'll properly correct it soon. `type` means `actor type`, it's the resource docker performs an action with. A question that this interpretation raises is if the type should then be inside the `Actor` structure, that's sounds reasonable, and I'm okay either way.

Docker performs actions over the resources that it manages, therefore, what you consider types are labeled as `action`. Those actions an logged in the event log, but again, the events do NOT describe themselves, they describe the actions that docker performs.

**Comment by thaJeztah** [comment]
> That could be an interpretation, I'm going to give you another one
> Docker performs actions over the resources that it manages, therefore, what you consider types are labeled as `action`

Guess that makes sense as well. Alright, let's stick with your implementation, "+1" on design for me.

File: daemon/container_operations_unix.go [hunk 3]
```diff
@@ -718,6 +722,11 @@ func (daemon *Daemon) DisconnectFromNetwork(container *container.Container, n li
 	if err := container.ToDiskLocking(); err != nil {
 		return fmt.Errorf("Error saving container to disk: %v", err)
 	}
+
+	attributes := map[string]string{
+		"container": container.ID,
+	}
+	daemon.LogNetworkEventWithAttributes(n, "disconnect", attributes)
 	return nil
 }
```
Comment (aboch) [inline daemon/container_operations_unix.go]: In this file can you please also take care of
`func (daemon *Daemon) releaseNetwork(container *Container)`

Given it is for a forced sandbox removal, I think we should generate the disconnect event from each connected network with something like:
```
+	attributes := map[string]string{
+		"container": container.ID,
+	}
+
 	for n := range networks {
+		if nw, err := daemon.FindNetwork(n); err == nil {
+                     daemon.LogNetworkEventWithAttributes(n, "disconnect", attributes)
+		}
 		networks[n] = &network.EndpointSettings{}
 	}
```

Comment (calavera) [inline daemon/container_operations_unix.go]: sounds good, thanks.

Comment (calavera) [inline daemon/container_operations_unix.go]: @aboch I think it's very weird to raise a disconnection event here when the container has not even connected to the network yet. As far as I understand this code, this is only a precaution to cleanup the sanbox before the container connects to a network, isn't it?

Comment (aboch) [inline daemon/container_operations_unix.go]: @calavera Actually the `releaseNetwork()` is called by the monitor process every time the container stops.
And also on container start failure, but only if the network connect was successful.

In either case the logging would happen only for the networks the container was actually connected to.

Comment (calavera) [inline daemon/container_operations_unix.go]: yeah, the fact that the caller is in `daemon/start.go` always confuses me. Thank you.

**Comment by icecrime** [comment]
```
FAIL: docker_cli_events_unix_test.go:259: DockerSuite.TestEventsFilterVolumeAndNetworkType

docker_cli_events_unix_test.go:269:
    c.Assert(events, checker.HasLen, 2)
... obtained []string = []string{"2015-12-30T05:38:48.136240021Z network create a8dd34c70483e89271d405328194956131dbf92b109c291cbc42945d9bff8bf1 (name=test-event-network-local, type=bridge)", "2015-12-30T05:38:48.326981430Z network destroy a8dd34c70483e89271d405328194956131dbf92b109c291cbc42945d9bff8bf1 (name=test-event-network-local, type=bridge)", "2015-12-30T05:38:49.134810430Z network create 940a7b517e2d70a719ac6cc1c9954bade533e54d80487bcbcaaaf2ed9f41b0fd (type=bridge, name=test-event-network-type)", "2015-12-30T05:38:49.163590109Z volume create test-event-volume-type (driver=local)"}
... n int = 2
```

**Comment by thaJeztah** [comment]
Looking at the discussion above, I think we're ready to move this to code review (and fix the failing tests during that)

**Comment by duglin** [comment]
Just now getting to look at this, sorry for the delay. 
SGTM
As for the order, I prefer what is proposed - resource type being acted upon first, then the action. I think starting with something more abstract and getting more specific as we move to the right in the list is the easiest to deal with.  Of course, then actor_id should come before the action but the problem with that is its then a little harder to parse/grep when you just want to find "network create" it would need to be "network .* create " - not a huge issue but the human readability is better this way.

**Comment by jfrazelle** [comment]
LGTM

**Comment by prologic** [comment]
Will this break tools that rely on the current structure events from the Docker Events API such as [autodock](https://github.com/prologic/autodock) which uses [docker-py](https://pypi.python.org/pypi/docker-py)?

**Comment by calavera** [comment]
@prologic I don't think there is a solid "yes" or "no" answer to that, it depends on how your project uses the events api. It might probably need some adjustments. Check what I wrote in the description about backwards compatibility:

image

Let me know if you have further questions about it.

**Comment by prologic** [comment]
@calavera Thanks for the clarification! I'll keep an eye out for this one as autodock is a consumer of the Docker Events API. I've already had to adjust for new events in Docker 1.9 as well as some internal structural changes in the data from inspecting container configurations. Thanks again!

**Comment by jfrazelle** [comment]
LGTM

**Comment by LK4D4** [comment]
LGTM

**Comment by moxiegirl** [comment]
Was it likely customers see any visible changes in output --- for example if they were crunching output from the API --- do we need to mention something as being deprecated?

Otherwise, LGTM.

**Comment by calavera** [comment]
good point @moxiegirl. The API still works with old fields, but I agree that's worth noticing in the deprecation document. I added it already.

**Comment by thaJeztah** [comment]
docs LGTM

@moxiegirl the API is backwards compatible, but CLI output changed, I think we should mention that in the release notes

**Comment by moxiegirl** [comment]
@thaJeztah  If by release notes you mean the changelog. Then we are good. :-)

**Comment by thaJeztah** [comment]
potatoes potatos, yup that's what I meant, haha

**Comment by TomasTomecek** [comment]
@calavera very nice! can't wait to play with this

**Comment by thaJeztah** [comment]
@TomasTomecek for playing; https://master.dockerproject.org is your friend :-) And please do! We can never have enough testers :)

**Pull request merged**